hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java

```diff
@@ -486,6 +491,57 @@ public void makeQualified(FileStatus status, URI uri, Path path,
     return new IteratorAdapter(bucket.listKeys(key));
   }
 
+  /**
+   * Helper for OFS listStatus on root.
+   */
+  private List<FileStatusAdapter> listStatusRoot(
+      boolean recursive, String startPath, long numEntries,
+      URI uri, Path workingDir, String username) throws IOException {
+    OFSPath ofsStartPath = new OFSPath(startPath);
+    // list volumes for user
+    Iterator<? extends OzoneVolume> iter = objectStore.listVolumesByUser(
+        username, null, ofsStartPath.getVolumeName());
+    List<FileStatusAdapter> res = new ArrayList<>();
+    // TODO: Test continuation
+    while (iter.hasNext() && res.size() <= numEntries) {
+      OzoneVolume volume = iter.next();
+      res.add(getFileStatusAdapterForVolume(volume, uri));
+      if (recursive) {
+        String next = volume.getName();
+        // TODO: Check startPath
+        res.addAll(listStatus(next, recursive, startPath,
+            numEntries - res.size(), uri, workingDir, username));
+      }
+    }
+    return res;
+  }
+
+  /**
+   * Helper for OFS listStatus on a volume.
+   */
+  private List<FileStatusAdapter> listStatusVolume(String volumeStr,
+      boolean recursive, String startPath, long numEntries,
+      URI uri, Path workingDir, String username) throws IOException {
+    OFSPath ofsStartPath = new OFSPath(startPath);
```

Comment:
should we use startBucket as parameter to avoid creating OFSPath unnecessarily?

Reply:
I think we can implement this with either next or startPath but not necessarily both.

Reply:
> should we use startBucket as parameter to avoid creating OFSPath unnecessarily?

Done in 41c081f2f6e506f0942fe89ad5ced91df6e2c3d0

Reply:
> I think we can implement this with either next or startPath but not necessarily both.

I just realized that the current implementation with both continuation (startPath) and recursion might not work as expected.

e.g. if such call is made: `listStatus(pathStr="/", recursive=true, startPath="/volume2/bucket2/dir2")`, the result won't be what we expect.